internal/spec/parser.go

```diff
@@ -242,11 +250,16 @@ func (p interfaceMethodParser) extractModelOrSliceReturns(returns *types.Tuple)
 	if !types.Identical(returns.At(1).Type(), code.TypeError) {
 		return "", NewUnsupportedReturnError(returns.At(1).Type(), 1)
 	}
-
+	isSamePkg := returns.At(0).Pkg() == p.NamedStruct.Obj().Pkg()
 	switch t := returns.At(0).Type().(type) {
 	case *types.Pointer:
-		if types.Identical(t.Elem(), p.NamedStruct) {
+		if isSamePkg && types.Identical(t.Elem(), p.NamedStruct) {
 			return QueryModeOne, nil
+		} else {
+			// TODO actually compare with imported Named type
+			if t.Elem().String() == p.NamedStruct.String() {
```

Comment:
not sure if this is cutting corner 🤔

Reply:
thanks for sharing insightful detail! 

Since we expect to only have either 1 or 2 packages. 
in 7becd5afdac1c4595a7930e0b63ed9bf40cf6de4 I just use another reference to locate interface package :p

Reply:
I have an idea. What we want is a way to derive absolute package path from any kind of package path, so we may do something like this.

1. Derive model absolute path from cmdline args
2. Derive interface absolute path from cmdline args
3. Load packages with those derived absolute path and map package path to the result from packages.Load()

For derivation, we may need to call another 1 or 2 `packages.Load()`. But this time we don't need types, so we may left the first parameter of `packages.Load()` to be nil. The result of `Load` should return only 1 package. We will reject 0 package return (package not found) and >=2 package returns (in case of wildcard package paths `./abc/...`).

In total, we may need 2 to 3 `packages.Load()` call in order to derive absolute path and load types. I'm not sure if it will have a performance issue since `packages.Load()` is not that fast (even with nil config takes around 40ms per call), but you can share your thoughts!

Reply:
I agree that code would be cleaner with your propose.

I suppose it's something like this:

```go
func generateFromRequest(interfaceDir, interfaceName, modelDir, structModelName, destPkg string) (string, error) {
	cfg := packages.Config{
		Mode: packages.NeedName | packages.NeedTypes,
	}
	if modelDir == "" || modelDir == interfaceDir {
		pkgs, err := packages.Load(&cfg, interfaceDir)
		if err != nil {
			return "", err
		}
		if err := assertPackageCount(pkgs, 1, 1); err != nil {
			return "", err
		}
		return generateRepository(pkgs[0].Types, pkgs[0].Types, structModelName, interfaceName)
	}
	intfPkgID, err := getPkgID(interfaceDir)
	if err != nil {
		return "", err
	}
	modelPkgID, err := getPkgID(modelDir)
	if err != nil {
		return "", err
	}
	patterns := []string{intfPkgID, modelPkgID}
	pkgs, err := packages.Load(&cfg, patterns...)
	if err != nil {
		return "", err
	}
	pkgM := toMap(pkgs)
	if err := assertPackageCount(pkgs, 1, 2); err != nil {
		return "", err
	}
	switch len(pkgs) {
	case 1:
		// model and interface in the same package
		return generateRepository(pkgs[0].Types, pkgs[0].Types, structModelName, interfaceName)
	case 2:
		return generateRepository(pkgM[intfPkgID].Types, pkgM[modelPkgID].Types, structModelName, interfaceName)
	}
	return "", errors.New("unexpected error")
}

func getPkgID(pattern string) (string, error) {
	pkgs, err := packages.Load(nil, pattern)
	if err != nil {
		return "", err
	}
	if err := assertPackageCount(pkgs, 1, 1); err != nil {
		return "", err
	}
	return pkgs[0].ID, nil
}

func assertPackageCount(pkgs []*packages.Package, min, max int) error {
	if len(pkgs) < min {
		return errNoPackageFound
	}
	if len(pkgs) > max {
		return errUnsupportMultiplePkgs
	}
	return nil
}

func toMap(pkgs []*packages.Package) map[string]*packages.Package {
	m := make(map[string]*packages.Package)
	for _, pkg := range pkgs {
		m[pkg.ID] = pkg
	}
	return m
}
```

Reply:
For the performance concern, I've tested [the code](https://github.com/sunboyy/repogen/pull/49#discussion_r1605680328) with the 593c9748149de92c4e2ca67f7e0a78f2593103ff benchmark on my laptop.

I am not quite sure about the outcome since even with 2 more `packages.Load` ~80ms would not contribute to such huge difference ( 133 ms v.s. 795 ms). 🤔 

```
goos: darwin
goarch: arm64
pkg: github.com/sunboyy/repogen
BenchmarkLoadPkg1
iteration 1 , time cost: 37.005167ms
iteration 31 , time cost: 24.079243ms
iteration 49 , time cost: 24.037926ms
BenchmarkLoadPkg1-10                                	      49	  24038301 ns/op
BenchmarkLoadPkg2
iteration 1 , time cost: 24.510166ms
iteration 48 , time cost: 25.37205ms
BenchmarkLoadPkg2-10                                	      48	  25372506 ns/op
BenchmarkGenerateFromRequestGettingStartedNew
2024/05/18 13:25:45 Generating method: CountByCity
...
iteration 1 , time cost: 127.472125ms
2024/05/18 13:25:45 Generating method: CountByCity
...

iteration 8 , time cost: 133.921411ms
BenchmarkGenerateFromRequestGettingStartedNew-10    	       8	 133922505 ns/op
BenchmarkGenerateFromRequestCrossPkg
2024/05/18 13:25:47 Generating method: CountByCity
...
iteration 1 , time cost: 815.89825ms
2024/05/18 13:25:47 Generating method: CountByCity
...
iteration 2 , time cost: 795.237625ms
BenchmarkGenerateFromRequestCrossPkg-10             	       2	 795249521 ns/op
PASS
ok  	github.com/sunboyy/repogen	7.043s
```

here is the reference on main branch and is consistent with `BenchmarkGenerateFromRequestGettingStartedNew`
```
$ go test -v -bench=BenchmarkGenerateFromRequestGettingStarted -run=none .
goos: darwin
goarch: arm64
pkg: github.com/sunboyy/repogen
BenchmarkGenerateFromRequestGettingStarted
2024/05/18 13:28:59 Generating method: CountByCity
...
iteration 1 , time cost: 138.597041ms
2024/05/18 13:28:59 Generating method: CountByCity
...
iteration 8 , time cost: 130.001296ms
BenchmarkGenerateFromRequestGettingStarted-10    	       8	 130001927 ns/op
PASS
ok  	github.com/sunboyy/repogen	1.317s
```

Reply:
The code looks nice! But I would like to request a few improvements. Here is the draft:

```go
absoluteModelPkgID, err := getPkgID(...)
// assert err
absoluteRepoPkgID, err := getPkgID(...)
// assert err

// cfg := ...
pkgs, err := packages.Load(cfg, absoluteModelPkgID, absoluteRepoPkgID)
// assert err

// construct a map and get *types.Package instance by absoluteModelPkgID and absoluteRepoPkgID

return generateRepository(pkgM[intfPkgID].Types, pkgM[modelPkgID].Types, structModelName, interfaceName)
```

1. Package count assertion on the final `packages.Load` might not be necessary. Thus, you may do one of these options:
  - Remain `assertPackageCount`, but instead of checking for min/max, checking for equality only.
  - Rename `assertPackageCount` to `assertOnePackage` and do not receive any integer parameter. This is for checking 1 package case only.
  - Delete `assertPackageCount`, moving its content to `getPkgID`. (I prefer this one)
2. For the switch-case at the bottom of the function, I think we can simplify case 1 to use the same logic as case 2, reducing execution branch.
3. On the top section of the function which handles the case where `modelDir == interfaceDir`, we may use the same logic as below too. I understand that your code will be more performant, but simple code is prone to less bugs in the future. If performance is going to be a problem in the future, I will think about it later.

Reply:
Thank you for the response to the performance concern. I've just known that! I will take it into consideration later on.

Reply:
sure for the code review let's discuss details in the commits.

Btw do you want to keep the benchmark test?

Reply:
I think I don't need it right now.